ding/entry/serial_entry_bco.py

```diff
@@ -0,0 +1,288 @@
+    collector_env.seed(cfg.seed)
+    evaluator_env.seed(cfg.seed, dynamic_seed=False)
+    collector = create_serial_collector(
+        cfg.policy.collect.collector,
+        env=collector_env,
+        policy=policy.collect_mode,
+        tb_logger=tb_logger,
+        exp_name=cfg.exp_name
+    )
+    evaluator = InteractionSerialEvaluator(
+        cfg.policy.eval.evaluator, evaluator_env, policy.eval_mode, tb_logger, exp_name=cfg.exp_name
+    )
+    commander = BaseSerialCommander(
+        cfg.policy.other.commander, learner, collector, evaluator, None, policy=policy.command_mode
+    )
+    learned_model = InverseDynamicsModel(cfg.policy.model.obs_shape, cfg.policy.model.action_shape)
+    # ==========
+    # Main loop
+    # ==========
+    learner.call_hook('before_run')
+    end = time.time()
+    m = nn.Softmax(dim=-1)
+    collect_episode = copy.deepcopy(cfg.policy.collect.n_episode)
+    agent_data = list()
+    for epoch in range(cfg.policy.learn.train_epoch):
+        collect_kwargs = commander.step()
+        # Evaluate policy performance
+        if evaluator.should_eval(learner.train_iter):
+            stop, reward = evaluator.eval(learner.save_checkpoint, learner.train_iter, collector.envstep)
+            if stop:
+                break
+        new_data = collector.collect(
+            n_episode=collect_episode, train_iter=learner.train_iter, policy_kwargs=collect_kwargs
+        )
+        agent_data = agent_data + new_data
+        learn_dataset = load_agentdata(agent_data)
+        learn_dataloader = DataLoader(learn_dataset, cfg.policy.learn.batch_size, sampler=None, num_workers=3)
+        for i, train_data in enumerate(learn_dataloader):
+            learned_model = train_state_trainsition_model(train_data, learned_model, 10)
+        # Generate state transitions from demonstrated state trajectories by IDM
+        expert_action_data = torch.argmax(m(learned_model.forward(expert_learn_dataset.obs)['logit']), -1)
+        post_expert_dataset = BCODataset(
+            {
+                'obs': expert_learn_dataset.obs[:, 0:int(expert_learn_dataset.obs.shape[1] / 2)],
+                'action': expert_action_data
+            }
+        )  # post_expert_dataset: Only obs and action are reserved for BC. next_obs are deleted
+        expert_learn_dataloader = DataLoader(
+            post_expert_dataset, cfg.policy.learn.batch_size, sampler=None, num_workers=3
+        )
+        # Improve policy using BC
+        for i, train_data in enumerate(expert_learn_dataloader):
+            learner.data_time = time.time() - end
+            learner.epoch_info = (epoch, i, len(learn_dataloader))
+            learner.train(train_data, collector.envstep)
+            end = time.time()
+        learner.policy.get_attribute('lr_scheduler').step()
+        collect_episode = int(cfg.policy.collect.n_episode * cfg.policy.collect.alpha)
```

Comment:
this variable should be calculated before this for loop

Reply:
this variable is different in first epoch(n_episode) and other epochs(n_episode*alpha).  so this is prepared for the next epoch. 
But it calculated repeatly. I modified it as a switch.